From 672ddeb3f70e008f6ce8c8925435c7f2d06b70e2 Mon Sep 17 00:00:00 2001
From: Au Dang <au.dang.jx@renesas.com>
Date: Wed, 26 Jun 2024 13:29:32 +0700
Subject: [PATCH 2/4] drivers: pwm: gpt-rzg2l: Add missing get reset for GPT

RZ/V2H needs 2 reset signals to work.
So, update driver to get the reset array instead of a single
reset controller.

Signed-off-by: Au Dang <au.dang.jx@renesas.com>
---
 drivers/pwm/gpt-rzg2l.c | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/drivers/pwm/gpt-rzg2l.c b/drivers/pwm/gpt-rzg2l.c
index 64a4943579bc..bb03dde37ff0 100644
--- a/drivers/pwm/gpt-rzg2l.c
+++ b/drivers/pwm/gpt-rzg2l.c
@@ -2295,6 +2295,11 @@ static const struct attribute_group buffer_attr_group = {
 	.attrs = buffer_attrs,
 };
 
+static void gpt_reset_control_assert(void *data)
+{
+	reset_control_assert(data);
+}
+
 static int rzg2l_gpt_probe(struct platform_device *pdev)
 {
 	struct rzg2l_gpt_chip *rzg2l_gpt;
@@ -2374,13 +2379,24 @@ static int rzg2l_gpt_probe(struct platform_device *pdev)
 		return ret;
 	}
 
-	rzg2l_gpt->rstc = devm_reset_control_get_shared(&pdev->dev, NULL);
+	rzg2l_gpt->rstc = devm_reset_control_array_get(&pdev->dev, true, false);
 	if (IS_ERR(rzg2l_gpt->rstc)) {
 		dev_err(&pdev->dev, "failed to get cpg reset\n");
 		return PTR_ERR(rzg2l_gpt->rstc);
 	}
 
-	reset_control_deassert(rzg2l_gpt->rstc);
+	ret = reset_control_deassert(rzg2l_gpt->rstc);
+	if (ret) {
+		dev_err(&pdev->dev, "failed to deassert reset %d\n", ret);
+		return ret;
+	}
+
+	ret = devm_add_action_or_reset(&pdev->dev, gpt_reset_control_assert,
+				       rzg2l_gpt->rstc);
+	if (ret) {
+		dev_err(&pdev->dev, "failed to register assert devm action, %d\n", ret);
+		return ret;
+	}
 
 	irq = platform_get_irq_byname(pdev, "gtcib");
 	if (irq < 0) {
-- 
2.25.1

